File: 백준/Platinum/12899. 데이터 구조/데이터 구조.py
import sys
from math import ceil,log2
input = sys.stdin.readline

n = int(input())
p=2000000
N = 2**ceil(log2(p))

seg = [0]*(N*2+1)


def modify(list,num,v):
    index = N + num
    while index>0:
        list[index]+=v
        index>>=1

def search(list,goal):
    index = 1
    while index < N:
        l,r = index<<1, (index<<1) + 1
        if list[l]>=goal:
            index = l
        else:
            goal -= list[l]
            index = r
    return index - N


for i in range(n):
    a,b = map(int,input().split())
    if a==1:
        modify(seg,b,1)
    else:
        c = search(seg,b)
        print(c)
        modify(seg,c,-1)

File: 백준/Platinum/12899. 데이터 구조/데이터 구조.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int[] seg;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine());

        int p = 2000000;
        N = 1;
        while (N < p) N <<= 1;  // N = 2^ceil(log2(p))

        seg = new int[N * 2];

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (a == 1) {
                modify(b, 1);
            } else {
                int c = search(b);
                sb.append(c).append("\n");
                modify(c, -1);
            }
        }

        System.out.print(sb);
    }

    // 세그먼트 트리 수정
    static void modify(int num, int v) {
        int index = N + num;
        while (index > 0) {
            seg[index] += v;
            index >>= 1;
        }
    }

    // 세그먼트 트리 k번째 수 찾기
    static int search(int goal) {
        int index = 1;
        while (index < N) {
            int l = index << 1;
            int r = l + 1;
            if (seg[l] >= goal) {
                index = l;
            } else {
                goal -= seg[l];
                index = r;
            }
        }
        return index - N;
    }
}
